runtime/signals.c

```diff
@@ -187,8 +189,10 @@ CAMLexport void caml_leave_blocking_section(void)
      examined by [caml_process_pending_signals_exn], then
      [signals_are_pending] is 0 but the signal needs to be
      handled at this point. */
-  signals_are_pending = 1;
-  //caml_raise_if_exception(caml_process_pending_signals_exn());
+  if (check_for_pending_signals()) {
+    signals_are_pending = 1;
+    caml_set_action_pending();
+  }
```

Comment:
LGTM. At #8961 I was simply calling `caml_set_action_pending` after removing `signals_are_pending`.

Reply:
I would prefer if we could call `caml_set_action_pending();` even if `check_for_pending_signals()` returns 0.

First, it does not cost much : the only cost is that we will examine all the potential callbacks at the next polling point. This is not much more expensive than calling `check_for_pending_signals()`. Second, with #9674, there will be another source of thread-local callback, which will need updating `caml_something_to_do` at every context switch.

Reply:
I'd prefer to keep it like this. The cost of `set_action_pending` isn't zero: the next allocation will fail and spill/restore all registers.

In many programs, entering and leaving a blocking section is massively more frequent than context-switching between threads. If you want to implement per-thread callbacks, I think it would make more sense to do them in caml_thread_{save,restore}_runtime_state in the systhreads library, which is called from the blocking_section_hooks. (Also, doing it there gives you the opportunity to optimise the common case where no context switch has occurred, and the thread taking the lock is the last one to release it)

Reply:
Okay, go for it.